lib/storage/src/content_manager/snapshots/recover.rs

```diff
@@ -135,47 +169,57 @@ pub async fn do_recover_from_snapshot(
                 .collect();
 
             if other_active_replicas.is_empty() {
-                // No other active replicas, we can activate this shard
-                // as there is no de-sync possible
-                if toc.is_distributed() {
-                    log::debug!(
-                        "Activating shard {} of collection {} with consensus",
-                        shard_id,
-                        collection_name
-                    );
-                    toc.send_set_replica_state_proposal(
-                        collection_name.to_string(),
-                        this_peer_id,
-                        *shard_id,
-                        ReplicaState::Active,
-                    )?;
-                } else {
-                    log::debug!(
-                        "Activating shard {} of collection {} locally",
-                        shard_id,
-                        collection_name
-                    );
-                    collection
-                        .set_shard_replica_state(*shard_id, this_peer_id, ReplicaState::Active)
-                        .await?;
-                }
+                activate_shard(toc, &collection, this_peer_id, shard_id).await?;
             } else {
-                let (replica_peer_id, _state) = other_active_replicas.into_iter().next().unwrap();
-                log::debug!(
-                    "Running synchronization for shard {} of collection {} from {}",
-                    shard_id,
-                    collection_name,
-                    replica_peer_id
-                );
-
-                // assume that if there is another peers, the server is distributed
-                toc.request_shard_transfer(
-                    collection_name.to_string(),
-                    *shard_id,
-                    *replica_peer_id,
-                    this_peer_id,
-                    true,
-                )?;
+                match priority {
+                    SnapshotPriority::Snapshot => {
+                        // Snapshot is the source of truth, we need to remove all other replicas
+                        activate_shard(toc, &collection, this_peer_id, shard_id).await?;
+
+                        let mut replicas_to_keep = state.config.params.replication_factor.get() - 1;
+
+                        for (peer_id, _) in other_active_replicas {
+                            if replicas_to_keep > 0 {
+                                // Keep this replica
+                                replicas_to_keep -= 1;
+
+                                toc.send_set_replica_state_proposal(
```

Comment:
Is the order correct? When we add a new replica first, we get the amount of replicas > replica factor. Otherwise, we can delete obsolete replica first - obsolete replica data is invalid and we don't lose user data.

Reply:
Close conversation, amount of replicas > `replica_factor` is fine